controllers/propagator/template_utils.go

```diff
@@ -0,0 +1,647 @@
+
+	wg.Add(1)
+
+	go func() {
+		log := log.WithValues("serviceAccount", saName, "namespace", namespace)
+
+		log.V(2).Info("Got a token", "expiration", expirationTimestamp.UTC().Format(time.RFC3339))
+
+		defer func() {
+			if err := os.Remove(tokenFilePath); err != nil {
+				log.Error(err, "Failed to clean up the service account token file")
+			}
+
+			wg.Done()
+		}()
+
+		// The latest refresh of the token is MinRefreshMins minutes before expiration
+		minRefreshBuffer := time.Duration(refreshConfig.MinRefreshMins * float64(time.Minute))
+		// Get the acceptable range of minutes that can be variable for the jitter below
+		maxJitterMins := (refreshConfig.MaxRefreshMins - refreshConfig.MinRefreshMins) * float64(time.Minute)
+
+		for {
+			// Add a jitter between 0 and maxJitterMins to make the token renewals variable
+			jitter := time.Duration(rand.Float64() * maxJitterMins).Round(time.Second) // #nosec G404
+			// Refresh the token between 5 and 10 minutes from now
+			refreshTime := expirationTimestamp.Add(-minRefreshBuffer).Add(-jitter)
+
+			log.V(2).Info("Waiting to refresh the token", "refreshTime", refreshTime)
+
+			deadlineCtx, deadlineCtxCancel := context.WithDeadline(ctx, refreshTime)
+
+			<-deadlineCtx.Done()
+
+			deadlineErr := deadlineCtx.Err()
+
+			// This really does nothing but this satisfies the linter that cancel is called.
+			deadlineCtxCancel()
+
+			if !errors.Is(deadlineErr, context.DeadlineExceeded) {
+				return
+			}
+
+			log.V(1).Info("Refreshing the token")
+
+			returnedTokenReq, err := client.CoreV1().ServiceAccounts(namespace).CreateToken(
+				ctx, saName, req, metav1.CreateOptions{},
+			)
+			if err != nil {
+				log.Error(err, "Failed to renew the service account token. Stopping the template resolver.")
+
+				if refreshConfig.OnFailedRefresh != nil {
+					refreshConfig.OnFailedRefresh(err)
+				}
+
+				return
+			}
+
+			err = os.WriteFile(tokenFilePath, []byte(returnedTokenReq.Status.Token), 0o600)
+			if err != nil {
+				log.Error(err, "The refreshed token couldn't be written to. Stopping the template resolver.")
+
+				if refreshConfig.OnFailedRefresh != nil {
+					refreshConfig.OnFailedRefresh(err)
+				}
+
+				return
+			}
+
+			expirationTimestamp = returnedTokenReq.Status.ExpirationTimestamp
+		}
+	}()
+
+	return tokenFilePath, nil
+}
+
+// getUserKubeConfig will copy the URL and TLS related configuration from `config` and set the input token file when
+// generating a new configuration. No authentication configuration is copied over.
+func getUserKubeConfig(config *rest.Config, tokenFile string) *rest.Config {
+	userConfig := &rest.Config{
+		Host:    config.Host,
+		APIPath: config.APIPath,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAFile:     config.TLSClientConfig.CAFile,
+			CAData:     config.TLSClientConfig.CAData,
+			ServerName: config.TLSClientConfig.ServerName,
+			// For testing
+			Insecure: config.TLSClientConfig.Insecure,
```

Comment:
This comment makes me think this piece should be removed, but I don't think that's right - this just means that this user config's Insecure setting matches the input rest Config's setting, which seems good even outside of testing.